src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -252,9 +320,38 @@ impl SkipWatermarkState {
                     continue;
                 }
                 Ordering::Equal => {
+                    self.last_serde = self
+                        .compaction_catalog_agent_ref
+                        .watermark_serde(table_id.table_id());
+
                     match direction {
                         WatermarkDirection::Ascending => {
-                            match inner_key.cmp(watermark.as_ref()) {
+                            let cmp = match watermark {
+                                WatermarkSerdeType::Bytes(watermark) => {
+                                    inner_key.cmp(watermark.as_ref())
+                                }
+                                WatermarkSerdeType::Serde(watermark) => {
+                                    let (
+                                        pk_prefix_serde,
+                                        watermark_col_serde,
+                                        watermark_col_idx_in_pk,
+                                    ) = self.last_serde.as_ref().unwrap();
+                                    let row = pk_prefix_serde
+                                            .deserialize(inner_key)
+                                            .unwrap_or_else(|_| {
+                                                panic!("Failed to deserialize pk_prefix inner_key {:?} serde data_types {:?} order_types {:?}", inner_key, pk_prefix_serde.get_data_types(), pk_prefix_serde.get_order_types());
+                                            });
+                                    let watermark_col_in_pk =
+                                        row.datum_at(*watermark_col_idx_in_pk);
+                                    cmp_datum(
```

Comment:
IIUC, if `cmp_datum` returns `Euqal | Greater`, based on the logic in L360, the watermark will be advanced. I think this is incorrect for non pk prefix watermark because the non pk prefix watermark and the pk doesn't have the same ordering.